lib/collection/src/shard/mod.rs

```diff
@@ -0,0 +1,87 @@
+pub mod local_shard;
+pub mod remote_shard;
+
+use crate::collection_manager::holders::segment_holder::SegmentHolder;
+use crate::shard::remote_shard::RemoteShard;
+use crate::{
+    CollectionInfo, CollectionResult, CollectionSearcher, CollectionUpdateOperations, LocalShard,
+    OptimizersConfigDiff, PointRequest, Record, SearchRequest, UpdateResult,
+};
+use async_trait::async_trait;
+use parking_lot::RwLock;
+use segment::types::{ExtendedPointId, Filter, ScoredPoint, WithPayload, WithPayloadInterface};
+use std::sync::Arc;
+use tokio::runtime::Handle;
+
+pub type ShardId = u32;
+
+pub type PeerId = u32;
+
+/// Shard
+///
+/// A shard can either be local or remote
+///
+#[allow(clippy::large_enum_variant)]
+pub enum Shard {
+    Local(LocalShard),
+    Remote(RemoteShard),
+}
+
+impl Shard {
+    pub fn get(&self) -> Arc<dyn ShardOperation + Sync + Send + '_> {
+        match self {
+            Shard::Local(local_shard) => Arc::new(local_shard),
+            Shard::Remote(remote_shard) => Arc::new(remote_shard),
+        }
+    }
+
+    pub async fn before_drop(&mut self) {
+        match self {
+            Shard::Local(local_shard) => local_shard.before_drop().await,
+            Shard::Remote(_) => (),
+        }
+    }
+}
+
+#[async_trait]
+pub trait ShardOperation {
+    async fn update(
+        &self,
+        operation: CollectionUpdateOperations,
+        wait: bool,
+    ) -> CollectionResult<UpdateResult>;
+
+    fn segments(&self) -> &RwLock<SegmentHolder>;
```

Comment:
is it possible for the remove shard?

Reply:
yeah, let's remove from the trait

Reply:
https://github.com/qdrant/qdrant/pull/384/commits/4a699d62bc3f1fb25fceb9b0a72f316c9e06c572